modules/engine/src/main/scala/com/gsk/kg/engine/optimizer/MovingForwardValues.scala

```diff
@@ -0,0 +1,82 @@
+package com.gsk.kg.engine
+package optimizer
+
+import cats.implicits._
+
+import higherkindness.droste.Basis
+
+import com.gsk.kg.engine.DAG._
+import com.gsk.kg.engine.data.ToTree._
+import com.gsk.kg.sparqlparser.Expr.Row
+import com.gsk.kg.sparqlparser.StringVal
+
+object MovingForwardValues {
+
+  def addJoinValues[T](
+      quads: List[T],
+      tableVars: List[StringVal.VARIABLE],
+      tableRows: List[Row]
+  )(implicit T: Basis[DAG, T]): BGP[T] = BGP(
+    quads.map(quad =>
+      quad match {
+        // tried to join the table with the quads whenever they have common variables.
+        case Quad(s, p, o, _)
+            if List(s, p, o).exists(elem => tableVars.contains(elem)) =>
+          val commonValues: List[StringVal.VARIABLE] =
+            tableVars.filter(List(s, p, o).contains(_))
+          if (commonValues == tableVars) {
+            T.algebra(Join(quad, T.algebra(Table(tableVars, tableRows))))
+          } else {
+            // checks whether rows are defined or undefined
+            val valueKeysOfRows: List[Seq[StringVal.VARIABLE]] =
+              tableRows.map(_.tuples.map { case (key, _) => key })
+            val allTheCommonValuesAreDef: Boolean = (for {
+              row     <- valueKeysOfRows
+              comdVal <- commonValues
+            } yield row.contains(comdVal)).reduce(_ && _)
+            if (allTheCommonValuesAreDef) {
+              // join only with the common and define values
+              val onlyCommonRows = tableRows.map(row =>
+                Row(row.tuples.filter { case (key, _) =>
+                  commonValues.contains(key)
+                })
+              )
+              T.algebra(
+                Join(quad, T.algebra(Table(commonValues, onlyCommonRows)))
+              )
+            } else {
+              quad
+            }
+          }
+        case _ => quad
+      }
+    )
+  )
+
+  def apply[T](implicit T: Basis[DAG, T]): T => T = { t =>
+    T.coalgebra(t).rewrite { case j @ Join(l, r) =>
+      (T.coalgebra(l), T.coalgebra(r)) match {
+        case (Table(tableVars, tableRows), BGP(quads)) =>
+          addJoinValues(quads, tableVars, tableRows)
+        case (BGP(quads), Table(tableVars, tableRows)) =>
+          addJoinValues(quads, tableVars, tableRows)
+        case _ => j
+      }
+    }
+  }
+
+  def phase[T](implicit T: Basis[DAG, T]): Phase[T, T] = Phase { t =>
+    val result = apply(T)(t)
+    (result != t)
+      .pure[M]
+      .ifM(
+        Log.debug(
+          "Optimizer(JoinValues)",
```

Comment:
What is the meaning of "Moving Forward" in this context? It seems unclear.

```suggestion
          "Optimizer(MovingForwardValues)",
```

Reply:
This change applies the join between the different quads and the table generated by the values VALUE clause, with common columns, instead of doing the join only once at the end of the process. The idea is to remove unwanted data as soon as possible. 

So it is a way to forward (and also distribute) the joins with the table generated in the VALUE clause in the DAG.

Perhaps a more convenient name could be advanceJoinValueClause

Reply:
So by "forward", do you mean "forward in time", as in "earlier"? If so, I think a name like "EarlyJoinOnValueClause" would be more clear. Please make sure to update the name in the debug statement also.